### PR TITLE
Improve concurrency in escalus_bosh (+fixes)

### PR DESCRIPTION
**Pull request by fenek**
This PR improves BOSH interleaving support in `escalus_bosh`. It ensures message order and tracks requests better.

### CODE REVIEW

File: src/escalus_bosh.erl [hunk 1]
```diff
@@ -504,3 +572,14 @@ host_to_list({_,_,_,_} = IP4) -> inet_parse:ntoa(IP4);
 host_to_list({_,_,_,_,_,_,_,_} = IP6) -> inet_parse:ntoa(IP6);
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space right "," on line 572

Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space right "," on line 572

Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space right "," on line 572

Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space right "," on line 572

Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space right "," on line 572

Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space right "," on line 572

Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Elvis**:
> Missing space right "," on line 572

File: src/escalus_bosh.erl [hunk 2]
```diff
@@ -82,7 +93,7 @@ connect(Args) ->
     {ok, Transport}.
 
 send(#client{rcv_pid = Pid} = Socket, Elem) ->
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Compiler**:
> Warning: missing specification for function send/2

File: src/escalus_bosh.erl [hunk 3]
```diff
@@ -241,6 +255,11 @@ recv(#client{rcv_pid = Pid}) ->
 get_requests(#client{rcv_pid = Pid}) ->
     gen_server:call(Pid, get_requests).
 
+%% This flag makes client to fail on stream error,
+%% even if it arrives out of order (according to RIDs)
+set_quickfail(#client{rcv_pid = Pid}, QuickfailFlag) ->
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Compiler**:
> Warning: missing specification for function set_quickfail/2

File: src/escalus_bosh.erl [hunk 4]
```diff
@@ -241,6 +255,11 @@ recv(#client{rcv_pid = Pid}) ->
 get_requests(#client{rcv_pid = Pid}) ->
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Compiler**:
> Warning: missing specification for function get_requests/1

File: src/escalus_bosh.erl [hunk 5]
```diff
@@ -336,20 +363,27 @@ handle_cast(reset_parser, #state{parser = Parser} = State) ->
...
-          of
-              {streamend, _, _} -> close_requests(NS#state{terminated=true});
-              {_, false, _}     -> NS;
-              {_, true, true}   -> send(Transport,
-                                        empty_body(NS#state.rid, NS#state.sid),
-                                        NS);
-              {_, true, false}  -> NS
-    end,
-    {noreply, NNS};
+handle_info(_, #state{ terminated = true } = S) ->
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Compiler**:
> Warning: missing specification for function handle_info/2

File: src/escalus_bosh.erl [hunk 6]
```diff
@@ -82,7 +93,7 @@ connect(Args) ->
     {ok, Transport}.
 
 send(#client{rcv_pid = Pid} = Socket, Elem) ->
-    gen_server:cast(Pid, {send, Socket, Elem}).
+    gen_server:call(Pid, {send, Socket, Elem}).
 
 is_connected(#client{rcv_pid = Pid}) ->
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Compiler**:
> Warning: missing specification for function is_connected/1

File: src/escalus_bosh.erl [hunk 7]
```diff
@@ -304,27 +329,29 @@ handle_call(recv, _From, State) ->
...
+handle_call(stop, From, #state{ waiting_requesters = WaitingRequesters } = State) ->
     StreamEnd = escalus_stanza:stream_end(),
-    {ok, _Reply, NewState} =
-    sync_send0(transport(State), exml:to_iolist(StreamEnd), State),
-    {stop, normal, ok, NewState}.
+    Ref = make_ref(),
+    NewState = wrap_and_send(transport(State), exml:to_iolist(StreamEnd), Ref, State),
+    {noreply, NewState#state{ waiting_requesters = [{Ref, From} | WaitingRequesters] }}.
 
 handle_cast(stop, State) ->
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Compiler**:
> Warning: missing specification for function handle_cast/2

File: src/escalus_bosh.erl [hunk 8]
```diff
@@ -270,11 +289,17 @@ init([Args, Owner]) ->
...
                 keepalive = proplists:get_value(keepalive, Args, true),
                 wait = Wait,
+                requests = queue:new(),
+                pending_requests = queue:new(),
                 event_client = EventClient,
                 client = Client,
                 on_reply = OnReplyFun}}.
 
 
+handle_call({send, Transport, Elem}, _From, State) ->
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Compiler**:
> Warning: missing specification for function handle_call/3

File: src/escalus_bosh.erl [hunk 9]
```diff
@@ -97,7 +108,10 @@ stop(#client{rcv_pid = Pid}) ->
...
             already_stopped;
         exit:{normal, {gen_server, call, _}} ->
-            already_stopped
+            already_stopped;
+        exit:{timeout, {gen_server, call, _}} ->
+            error({timeout, process_info(Pid, current_stacktrace),
+                   process_info(Pid, messages), catch sys:get_state(Pid)})
     end.
 
 kill(#client{} = Client) ->
```
Comment (gadgetci) [inline src/escalus_bosh.erl]: According to **Compiler**:
> Warning: missing specification for function kill/1

**Comment by fenek** [comment]
Let's wait for respective PR in MongooseIM to pass as well.

**Comment by fenek** [comment]
Build successful: https://github.com/esl/MongooseIM/pull/1289

**Comment by michalwski** [comment]
Thanks for the change! Now it's done the right way.

**Pull request merged**